File: backend/stub.py
"""Official payslip stub fields vs calculated payroll."""

from __future__ import annotations

from typing import Any

STUB_MONEY_KEYS = [
    "zakladna",
    "dovolenka",
    "osobne",
    "sviatky",
    "premie",
    "noc",
    "sobota",
    "nedela",
    "platene_volno",
    "premie_dlhsie",
    "hruba",
    "np",
    "sp",
    "ip",
    "pvn",
    "zp",
    "nczd",
    "dan",
    "danovy_bonus",
    "cista",
    "nahrada_prijmu",
    "nezdane_nahrady",
    "vyuctovanie",
    "er_np",
    "er_sp",
    "er_ip",
    "er_pvn",
    "er_pfp",
    "er_up",
    "er_gp",
    "er_prfs",
    "er_zp",
    "employer_cost",
]

STUB_QTY_KEYS = [
    "dovolenka_days",
    "sviatky_days",
    "noc_hours",
    "sobota_hours",
    "nedela_hours",
    "platene_volno_days",
]

# Čistá / vyúčtovanie count as the received amount, not as "details".
RECEIVED_STUB_KEYS = ("cista", "vyuctovanie")
DETAIL_MONEY_KEYS = [k for k in STUB_MONEY_KEYS if k not in RECEIVED_STUB_KEYS]

RECONCILE_ROWS = [
    ("Základná mzda", "basic", "zakladna"),
    ("Osobné ohodnotenie", "osobne", "osobne"),
    ("Príplatok sobota", "sat_prem", "sobota"),
    ("Príplatok nedeľa", "sun_prem", "nedela"),
    ("Práca v noci", "night_prem", "noc"),
    ("Príplatok sviatok", "holiday_prem", "sviatky"),
    ("Príplatok nadčas", "ot_prem", None),
    ("Hrubá mzda", "hruba", "hruba"),
    ("Nemocenské (NP)", "np", "np"),
    ("Starobné (SP)", "sp", "sp"),
    ("Invalidné (IP)", "ip", "ip"),
    ("Poist. v nezam. (PvN)", "pvn", "pvn"),
    ("Zdravotné (ZP)", "zp", "zp"),
    ("Nezdaniteľná časť", "nczd_applied", "nczd"),
    ("Daň", "dan", "dan"),
    ("Čistá mzda", "cista", "cista"),
    ("Celková cena práce", "employer_cost", "employer_cost"),
]

STUB_ONLY_ROWS = [
    ("Dovolenka", "dovolenka", "dovolenka_days"),
    ("Sviatky (days on stub)", "sviatky", "sviatky_days"),
    ("Prémie", "premie", None),
    ("Prémie za dlhšie obdobie", "premie_dlhsie", None),
    ("Platené voľno", "platene_volno", "platene_volno_days"),
    ("Daňový bonus - deti", "danovy_bonus", None),
    ("Náhrada príjmu", "nahrada_prijmu", None),
    ("Nezdaniteľné náhrady", "nezdane_nahrady", None),
    ("Vyúčtovanie", "vyuctovanie", None),
    ("Poistné zamestnávateľa NP", "er_np", None),
    ("Poistné zamestnávateľa SP", "er_sp", None),
    ("Poistné zamestnávateľa IP", "er_ip", None),
    ("Poistné zamestnávateľa PvN", "er_pvn", None),
    ("Poistné zamestnávateľa PFP", "er_pfp", None),
    ("Poistné zamestnávateľa UP", "er_up", None),
    ("Poistné zamestnávateľa GP", "er_gp", None),
    ("Poistné zamestnávateľa PRFS", "er_prfs", None),
    ("Poistné zamestnávateľa ZP", "er_zp", None),
]


def empty_stub() -> dict[str, Any]:
    out: dict[str, Any] = {}
    for key in STUB_MONEY_KEYS + STUB_QTY_KEYS:
        out[key] = None
    return out


def _num(value: Any) -> float | None:
    if value in (None, "", False):
        return None
    try:
        n = float(value)
    except (TypeError, ValueError):
        return None
    if n == 0:
        return 0.0
    return n


def normalize_stub(raw: Any) -> dict[str, Any]:
    src = raw if isinstance(raw, dict) else {}
    out = empty_stub()
    for key in STUB_MONEY_KEYS + STUB_QTY_KEYS:
        if key in src:
            out[key] = _num(src[key])
    return out


def stub_has_values(stub: dict[str, Any] | None) -> bool:
    if not stub:
        return False
    return any(v not in (None, 0, 0.0) for v in stub.values())


def compact_stub(stub: dict[str, Any] | None) -> dict[str, Any] | None:
    norm = normalize_stub(stub)
    kept = {k: v for k, v in norm.items() if v not in (None,)}
    return kept or None


def _nonzero(value: Any) -> bool:
    n = _num(value)
    return n is not None and n != 0


def received_from_entry(entry: Any) -> float | None:
    if not isinstance(entry, dict):
        return _num(entry) if entry not in (0, 0.0) else None
    stub = normalize_stub(entry.get("stub"))
    for key in ("vyuctovanie", "cista"):
        n = _num(stub.get(key))
        if n not in (None, 0.0):
            return n
    amount = entry.get("amount")
    n = _num(amount)
    if n in (None, 0.0):
        return None
    return n


def stub_incomplete(entry: Any, received: float | None) -> bool:
    if received is None:
        return False
    if not isinstance(entry, dict):
        return True
    stub = normalize_stub(entry.get("stub"))
    if any(_nonzero(stub.get(k)) for k in DETAIL_MONEY_KEYS):
        return False
    return True


def reconcile_month(payroll: dict[str, Any], stub: dict[str, Any] | None) -> dict[str, Any]:
    stub_n = normalize_stub(stub)
    rows = []
    for label, calc_key, stub_key in RECONCILE_ROWS:
        calc = _num(payroll.get(calc_key)) if calc_key else None
        stub_v = _num(stub_n.get(stub_key)) if stub_key else None
        delta = None
        if calc is not None and stub_v is not None:
            delta = round(stub_v - calc, 2)
        rows.append({"label": label, "calc": calc, "stub": stub_v, "delta": delta, "mapped": stub_key is not None})
    extras = []
    for label, stub_key, qty_key in STUB_ONLY_ROWS:
        stub_v = _num(stub_n.get(stub_key))
        qty = _num(stub_n.get(qty_key)) if qty_key else None
        if stub_v is None and qty is None:
            continue
        extras.append({"label": label, "stub": stub_v, "qty": qty})
    return {"rows": rows, "extras": extras}


def explain_month(payroll: dict[str, Any], stub: dict[str, Any] | None, incomplete: bool) -> list[str]:
    notes: list[str] = []
    if incomplete:
        notes.append("Only the received (čistá) amount is filled in. Add the rest of the payslip lines to compare against the calculation.")
    stub_n = normalize_stub(stub)
    if not stub_has_values(stub_n) and not incomplete:
        return notes
    stub_osobne = _num(stub_n.get("osobne")) or 0.0
    calc_osobne = _num(payroll.get("osobne")) or 0.0
    stub_prems = sum(_num(stub_n.get(k)) or 0.0 for k in ("noc", "sobota", "nedela", "sviatky"))
    calc_extra = (payroll.get("ot_prem") or 0) + (payroll.get("night_prem") or 0) + (payroll.get("sat_prem") or 0) + (payroll.get("sun_prem") or 0)
    if stub_osobne > calc_osobne + 20 and stub_prems < 0.05 and calc_extra > 0.05:
        notes.append(
            "Stub osobné is much higher than calculated osobné, and stub príplatky are ~0 while the calc has night/OT. Extra hours were likely booked as osobné on the stub — keep calculated osobné at 0 so it is not double-counted."
        )
    recon = reconcile_month(payroll, stub_n)
    for row in recon["rows"]:
        if row["delta"] is None:
            continue
        if abs(row["delta"]) < 0.05:
            continue
        notes.append(
            f"{row['label']}: calc {row['calc']:.2f} € vs stub {row['stub']:.2f} € ({row['delta']:+.2f} €)."
        )
    return notes

File: frontend/src/stub.js
export const STUB_GROUPS = [
  {
    title: "Earnings",
    fields: [
      { key: "zakladna", label: "Základná mzda", kind: "money" },
      { key: "dovolenka_days", label: "Dovolenka (days)", kind: "qty" },
      { key: "dovolenka", label: "Dovolenka", kind: "money" },
      { key: "osobne", label: "Osobné ohodnotenie", kind: "money" },
      { key: "sviatky_days", label: "Sviatky (days)", kind: "qty" },
      { key: "sviatky", label: "Sviatky", kind: "money" },
      { key: "premie", label: "Prémie", kind: "money" },
      { key: "noc_hours", label: "Práca v noci (h)", kind: "qty" },
      { key: "noc", label: "Práca v noci", kind: "money" },
      { key: "sobota_hours", label: "Práca v sobotu (h)", kind: "qty" },
      { key: "sobota", label: "Práca v sobotu", kind: "money" },
      { key: "nedela_hours", label: "Práca v nedeľu (h)", kind: "qty" },
      { key: "nedela", label: "Práca v nedeľu", kind: "money" },
      { key: "platene_volno_days", label: "Platené voľno (days)", kind: "qty" },
      { key: "platene_volno", label: "Platené voľno", kind: "money" },
      { key: "premie_dlhsie", label: "Prémie za dlhšie obdobie", kind: "money" },
      { key: "hruba", label: "Hrubá mzda", kind: "money" },
    ],
  },
  {
    title: "Deductions / net",
    fields: [
      { key: "np", label: "Nemocenské (NP)", kind: "money" },
      { key: "sp", label: "Starobné (SP)", kind: "money" },
      { key: "ip", label: "Invalidné (IP)", kind: "money" },
      { key: "pvn", label: "Poist. v nezam. (PvN)", kind: "money" },
      { key: "zp", label: "Zdravotné (ZP)", kind: "money" },
      { key: "nczd", label: "Nezdaniteľná časť", kind: "money" },
      { key: "dan", label: "Daň", kind: "money" },
      { key: "danovy_bonus", label: "Daňový bonus - deti", kind: "money" },
      { key: "cista", label: "Čistá mzda", kind: "money", receivedLike: true },
      { key: "nahrada_prijmu", label: "Náhrada príjmu", kind: "money" },
      { key: "nezdane_nahrady", label: "Nezdaniteľné náhrady", kind: "money" },
      { key: "vyuctovanie", label: "Vyúčtovanie", kind: "money", receivedLike: true },
    ],
  },
  {
    title: "Employer",
    fields: [
      { key: "er_np", label: "NP (employer)", kind: "money" },
      { key: "er_sp", label: "SP (employer)", kind: "money" },
      { key: "er_ip", label: "IP (employer)", kind: "money" },
      { key: "er_pvn", label: "PvN (employer)", kind: "money" },
      { key: "er_pfp", label: "PFP (employer)", kind: "money" },
      { key: "er_up", label: "UP (employer)", kind: "money" },
      { key: "er_gp", label: "GP (employer)", kind: "money" },
      { key: "er_prfs", label: "PRFS (employer)", kind: "money" },
      { key: "er_zp", label: "ZP (employer)", kind: "money" },
      { key: "employer_cost", label: "Celková cena práce", kind: "money" },
    ],
  },
];

export const STUB_KEYS = STUB_GROUPS.flatMap((g) => g.fields.map((f) => f.key));
const DETAIL_KEYS = STUB_GROUPS.flatMap((g) => g.fields.filter((f) => f.kind === "money" && !f.receivedLike).map((f) => f.key));

export const emptyStub = () => Object.fromEntries(STUB_KEYS.map((k) => [k, ""]));

export function stubFromMonth(month) {
  const src = month?.stub || {};
  const out = emptyStub();
  for (const key of STUB_KEYS) {
    if (src[key] != null && src[key] !== "") out[key] = src[key];
  }
  return out;
}

export function parseStub(stub) {
  const out = {};
  for (const key of STUB_KEYS) {
    const raw = stub?.[key];
    if (raw === "" || raw == null) continue;
    const n = Number(raw);
    if (!Number.isNaN(n)) out[key] = n;
  }
  return out;
}

export function fillStubFromCalc(stub, month) {
  const map = {
    zakladna: "basic",
    osobne: "osobne",
    sviatky: "holiday_prem",
    noc: "night_prem",
    noc_hours: "night_h",
    sobota: "sat_prem",
    sobota_hours: "sat_h",
    nedela: "sun_prem",
    nedela_hours: "sun_h",
    hruba: "hruba",
    np: "np",
    sp: "sp",
    ip: "ip",
    pvn: "pvn",
    zp: "zp",
    nczd: "nczd_applied",
    dan: "dan",
    cista: "cista",
    employer_cost: "employer_cost",
  };
  const out = { ...emptyStub(), ...stub };
  for (const [stubKey, calcKey] of Object.entries(map)) {
    const cur = stub?.[stubKey];
    if (cur !== "" && cur != null) continue;
    const v = month?.[calcKey];
    if (v == null || v === "") continue;
    out[stubKey] = v;
  }
  return out;
}

export function isStubIncomplete(received, stub) {
  if (received == null) return false;
  return !DETAIL_KEYS.some((key) => {
    const n = Number(stub?.[key]);
    return stub?.[key] !== "" && stub?.[key] != null && !Number.isNaN(n) && n !== 0;
  });
}

export const SETTINGS_GROUPS = [
  {
    title: "Identity",
    keys: ["name", "department", "employer", "personal_no", "health_insurer"],
  },
  {
    title: "Contract",
    keys: [
      "hourly_rate",
      "avg_earnings",
      "employment_type",
      "dohoda_type",
      "contract_h_week",
      "full_time_shift_hours",
      "unpaid_break_after",
      "unpaid_break_hours",
    ],
  },
  {
    title: "Tax",
    keys: ["apply_nczd", "nczd", "tax19", "tax25", "tax30", "tax35", "bracket19", "bracket25", "bracket30"],
  },
  {
    title: "Employee odvody",
    keys: ["apply_oop", "oop", "rate_np", "rate_sp", "rate_ip", "rate_pvn", "rate_zp"],
  },
  {
    title: "Employer odvody",
    keys: ["er_np", "er_sp", "er_ip", "er_pvn", "er_pfp", "er_up", "er_gp", "er_prfs", "er_zp"],
  },
  {
    title: "Príplatky",
    keys: ["prem_sat", "prem_sun", "prem_night", "prem_hol_pct", "prem_ot_pct", "min_wage_month", "min_wage_hour"],
  },
];

export const HIDDEN_SETTINGS = ["ot_enabled", "contract_hours_week"];

export function settingsGroups(draft) {
  const source = draft || {};
  return SETTINGS_GROUPS.map((group) => ({
    title: group.title,
    entries: group.keys.map((key) => [key, key in source ? source[key] : ""]),
  })).filter((group) => group.entries.length);
}
